ArduCopter/mode_autotune.cpp

```diff
@@ -1182,36 +1184,89 @@ inline bool Copter::ModeAutoTune::yaw_enabled() {
     return g.autotune_axis_bitmask & AUTOTUNE_AXIS_BITMASK_YAW;
 }
 
-// twitching_test - twitching tests
+// twitching_test_rate - twitching tests
 // update min and max and test for end conditions
-void Copter::ModeAutoTune::twitching_test(float measurement, float target, float &measurement_min, float &measurement_max)
+void Copter::ModeAutoTune::twitching_test_rate(float rate, float rate_target, float &meas_rate_min, float &meas_rate_max)
+{
+    // capture maximum rate
+    if (rate > meas_rate_max) {
+        // the measurement is continuing to increase without stopping
+        meas_rate_max = rate;
+        meas_rate_min = rate;
+    }
+
+    // capture minimum measurement after the measurement has peaked (aka "bounce back")
+    if ((rate < meas_rate_min) && (meas_rate_max > rate_target * 0.5f)) {
+        // the measurement is bouncing back
+        meas_rate_min = rate;
+    }
+
+    // calculate early stopping time based on the time it takes to get to 90%
+    if (meas_rate_max < rate_target * 0.75f) {
+        // the measurement not reached the 90% threshold yet
+        step_stop_time = step_start_time + (millis() - step_start_time) * 3.0f;
+        step_stop_time = MIN(step_stop_time, step_start_time + AUTOTUNE_TESTING_STEP_TIMEOUT_MS);
+    }
+
+    if (meas_rate_max > rate_target) {
+        // the measured rate has passed the target rate
+        step = UPDATE_GAINS;
+    }
+
+    if (meas_rate_max-meas_rate_min > meas_rate_max*g.autotune_aggressiveness) {
+        // the measurement has passed 50% of the target rate and bounce back is larger than the threshold
+        step = UPDATE_GAINS;
+    }
+
+    if (millis() >= step_stop_time) {
+        // we have passed the maximum stop time
+        step = UPDATE_GAINS;
+    }
+}
+
+// twitching_test_angle - twitching tests
+// update min and max and test for end conditions
+void Copter::ModeAutoTune::twitching_test_angle(float angle, float rate, float angle_target, float &meas_angle_min, float &meas_angle_max, float &meas_rate_min, float &meas_rate_max)
 {
     // capture maximum measurement
-    if (measurement > measurement_max) {
+    if (angle > meas_angle_max) {
+        // the measurement is continuing to increase without stopping
+        meas_angle_max = angle;
+        meas_angle_min = angle;
+    }
+
+    // capture maximum rate
+    if (rate > meas_rate_max) {
+        // the measurement is continuing to increase without stopping
+        meas_rate_max = rate;
+        meas_rate_min = rate;
+    }
+
+    // capture maximum rate
+    if (rate < meas_rate_min) {
         // the measurement is continuing to increase without stopping
```

Comment:
Both comments above seem wrong, as it's related to minimum.

Also, I don't know if it's on purpose, but I thought I should mention this check doesn't exist in `twitching_test_rate`.

Reply:
Thanks, I will correct them.